tests(refactor): Improve consistency and documentation for test helpers

georglauterbach opened the pull request:
# Description

Refactors test helper. Naming was incoherent and documentation was missing.

## Type of change

<!-- Delete options that are not relevant. -->

- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

polarathene commented on test/helper/common.bash:
```diff
@@ -117,131 +232,207 @@ function repeat_in_container_until_success_or_timeout() {
-    # +1 of starting count:
-    EXPECTED_COUNT=$(( $(__get_count) + 1 ))
-  fi
+  #
+  #                             +1 of starting count:
+  [[ -z ${EXPECTED_COUNT} ]] && EXPECTED_COUNT=$(bc <<< "$(__get_count) + 1")
```
I don't think the comment indentation is necessary?

I see you have reverted the increment logic here to an earlier attempt I had (_or this was written by you previously and applied without updating_), is this intentional?

I'm not fussed with either approach personally, but I don't find the `|| true`  in `__get_count()` with the workaround comment to be any better (_if anything more awkward?_). @casperklein might have some input if there is a reason to prefer it though.

casperklein replied:
For reference, explained here: https://github.com/docker-mailserver/docker-mailserver/pull/2997#discussion_r1070583876

From a logical point of view, `__get_count` shouldn't exit with code 1, if the function works as expected for our tests. With `bc`, this is not triggered, but the function still return with exit code 1. This is why I consider my solution as more accurate.

However, at the moment you two are primary working on/with the tests. That said, when you both are more comfortable with something else, I am fine with it 👍

georglauterbach replied:
Done

polarathene commented on test/helper/common.bash:
```diff
@@ -117,131 +232,207 @@ function repeat_in_container_until_success_or_timeout() {
+  _repeat_in_container_until_success_or_timeout \
+    "${TIMEOUT}" \
+    "${CONTAINER_NAME}" \
+    /bin/bash -c '[[ $(mailq) == *"Mail queue is empty"* ]]'
```
Being more explicit should be ok here. From what I've seen it's the only output from the command when valid.

```suggestion
    /bin/bash -c '[[ $(mailq) == "Mail queue is empty" ]]'
```

georglauterbach replied:
Done

polarathene commented on test/helper/setup.bash:
```diff
@@ -28,51 +31,58 @@ function __initialize_variables() {
-# but will be used in `common_container_create()` and implicitly in other helper methods.
-function init_with_defaults() {
+# in the file, then use `local TEST_TMP_CONFIG=$(_duplicate_config_for_container . "${UNIQUE_ID_HERE}")`
+function _init_with_defaults() {
   __initialize_variables
 
   export TEST_TMP_CONFIG
-  TEST_TMP_CONFIG=$(duplicate_config_for_container . "${CONTAINER_NAME}")
+
+  # In `setup_file()` the default name to use for the currently tested docker container
+  # is `${CONTAINER_NAME}` global defined here. It derives the name from the test filename:
+  # `basename` to ignore absolute dir path and file extension, only extract filename.
+  # In `setup_file()` creates a single copy of the test config folder to use for an entire test file:
+  TEST_TMP_CONFIG=$(_duplicate_config_for_container . "${CONTAINER_NAME}")
```
This looks like you applied this before the relevant PR was updated and merged. Comments appear to be outdated here.

georglauterbach replied:
Yep, sorry!

georglauterbach replied:
Done

polarathene commented on test/helper/common.bash:
```diff
@@ -117,131 +232,207 @@ function repeat_in_container_until_success_or_timeout() {
+function _reload_postfix() {
+  local CONTAINER_NAME=${1:-${CONTAINER_NAME}}
+
+  # Reloading Postfix config after modifying it in <2 sec will cause Postfix to delay, workaround that:
```
Since this PR is quite noisy in diff (_not fun with tracking info on changes with `git blame`_), it might be helpful to reference the PR that details the workaround reasoning?

```suggestion
  # Reloading Postfix config after modifying it within 2 seconds will cause Postfix to delay reading `main.cf`:
  # WORKAROUND: https://github.com/docker-mailserver/docker-mailserver/pull/2998
```

polarathene reviewed (changes requested):
I spotted a few minor concerns and suggestions, and one bug (_change to implicit container name in `mail_changedetector.bats`_). Otherwise looking  pretty good :+1:

casperklein commented on Makefile:
```diff
@@ -17,7 +17,7 @@ BATS_PARALLEL_JOBS     ?= 2
 all: lint build generate-accounts tests clean
 
-build:
+build: ALWAYS_RUN
```
fyi: https://github.com/docker-mailserver/docker-mailserver/pull/2857#discussion_r1013767738

georglauterbach commented on test/helper/common.bash:
```diff
@@ -1,49 +1,164 @@
+  local CONTAINER_NAME=${1:?Container name must be provided when using explicit}
   shift 1
-  run docker exec "${CONTAINER_NAME}" "${@}"
+  run _exec_in_container_explicit "${CONTAINER_NAME:?Container name must be provided}" "${@}"
```
Done

georglauterbach commented on test/tests/serial/mail_privacy.bats:
```diff
@@ -1,39 +1,39 @@
   assert_success
-  assert_output 0
+
+  _run_in_container_bash 'find /var/mail/localhost.localdomain/user1/new/ -maxdepth 1 -type f -printf "x\n"'
```
Done

georglauterbach commented on test/tests/serial/tests.bats:
```diff
@@ -298,88 +297,88 @@ EOF
-@test "checking accounts: user3 should have been added to /tmp/docker-mailserver/postfix-accounts.cf" {
-  docker exec mail /bin/sh -c "addmailuser user3@domain.tld mypassword"
+@test "accounts: user3 should have been added to /tmp/docker-mailserver/postfix-accounts.cf" {
+  _exec_in_container /bin/sh -c "addmailuser user3@domain.tld mypassword"
```
Done

georglauterbach commented on test/tests/serial/tests.bats:
```diff
@@ -23,32 +24,30 @@ setup_file() {
-  wait_until_change_detection_event_completes mail
-  wait_for_service mail postfix
-  wait_for_smtp_port_in_container mail
+  _wait_until_change_detection_event_completes mail
```
Done

georglauterbach commented:
I rewrote history so we can rebase-merge this. I incorporated the review feedback from @polarathene into the new commits. I think reviewing commit by commit is again a nice strategy, as this PR again is rather big.

---

What's new with the new commits? I added one nice feature: for some test helper functions, we can now omit the container name if `CONTAINER_NAME` is already set. I made sure this functionality is robust and added [a function](https://github.com/docker-mailserver/docker-mailserver/blob/8e0f126bcc5d7a6e90ab7f2ed19f758d93c4ad5a/test/helper/common.bash#L24-L62) that handles the naming. I also made sure documentation for this is on-point. It looks like this in VS Code for me:

![Screenshot from 2023-01-19 11-41-26](https://user-images.githubusercontent.com/44545919/213496327-18faa5a1-adc9-4462-a5f4-1ac9d5966ba0.png)

polarathene commented:
> I think reviewing commit by commit is again a nice strategy, as this PR again is rather big.

That's extremely tedious for us to review though with these kind of changes.

It takes a long time to go through each commit and compare line by line like I did last time to catch any errors. Each time you rewrite history like that it reduces the quality of the review as less effort will be put into repeating that process.

> I rewrote history so we can rebase-merge this. 

I don't see what warrants a merge other than squash in this PR?

I'd would have preferred to see new commits with requested changes so the diff of what actually changed was much smaller, then after approval by all means rewrite history then and I'll trust that's all it was.

---

> What's new with the new commits?

That sounds great, and I love the VSCode screenshot showing off the doc tooltip! :grinning: 

---

I'll start a review now, but like I said a thorough review is taxing, so I'm placing more trust in this time round that we don't have any accidents like in the initial review with testing the wrong container for a method :confounded: (_easy mistake to make, but hard to spot in large noisy/repetitive diffs_)

georglauterbach commented:
Yeah, again sorry for the force-push, but I think that squash-merging is not the way to go. If I am overruled, I will definitely squash-merge it though. Not a big deal :)

> **EDIT:** Praise Github, it actually offers a [diff of before / after the history rewrite](https://github.com/docker-mailserver/docker-mailserver/compare/1d268ca246b30d8d7cf0b21730e6d5cb16b6cadd..8e0f126bcc5d7a6e90ab7f2ed19f758d93c4ad5a) pray

I _definitely_ knew about that, thought you did too, haha :D That should do the trick and make the review easier, right?:)

polarathene commented on test/helper/common.bash:
```diff
@@ -1,247 +1,397 @@
-function wait_for_empty_mail_queue_in_container() {
-  local CONTAINER_NAME="${1}"
-  local TIMEOUT=${TEST_TIMEOUT_IN_SECONDS}
+# Checks whether the directory exists and then how many files it container.
```
Not sure if top-level is the right way to phrase it, but better to clarify the file check isn't recursive depth :sweat_smile: 

```suggestion
# Checks if the directory exists and then how many files it contains at the top-level.
```

polarathene commented on test/helper/common.bash:
```diff
@@ -1,247 +1,397 @@
+# @param ${3} = container name [OPTIONAL]
+function _count_files_in_directory_in_container()
+{
+  local DIRECTORY=${1:?No directoty provided}
```
```suggestion
  local DIRECTORY=${1:?No directory provided}
```

polarathene commented on test/helper/common.bash:
```diff
@@ -1,247 +1,397 @@
-    '=(debug|trace)$'
+# Adds a mail account and waits for the associated files to be created.
+#
+# @param ${1} = account name
```
Be consistent with the earlier parameter definition:

```suggestion
# @param ${1} = mail account name
```

polarathene commented on test/helper/common.bash:
```diff
@@ -1,247 +1,397 @@
+#
+# @param ${1} = service name
+# @param ${2} = container name [OPTIONAL]
+function _container_has_service_running() {
```
If a helper contains a `run` call, I think it should be more explicit on the function name with `_should` or `_run` prefix as a convention? It's ok to use `assert_` methods that don't depend on `run` context though.

```suggestion
function _should_have_service_running_in_container() {
```

Then again we have `_add_mail_account_then_wait_until_ready()` and that is more clearly an action and not something we'd use a `run` on when calling I think? So perhaps this is fine as-is :man_shrugging:

georglauterbach replied:
I think the name you proposed is better; I will update it.

polarathene reviewed (changes requested):
I also wanted to inquire about the changes that swap the `load` order to `setup,common` instead of `common,setup`? Was there a reason for that all? I had been using the latter as alphabetical sorting :sweat_smile:

polarathene commented:
> I _definitely_ knew about that, thought you did too, haha :D That should do the trick and make the review easier, right?:)

Yes, but it still took me about 2 hours to comb through.

I definitely would have missed a bunch if I had to wade through the full diff with old history interleaved/overwritten.

---

> Yeah, again sorry for the force-push, but I think that squash-merging is not the way to go.

I'm not entirely opposed to it, but I'm not sure what the benefit is? The master branch would have a single logical commit tied to this PR which has full discussion of review and commit history.

The changes are all related / dependent, so makes more sense to me for them to be bundled into a squash merge.

If this is about the two commits that refactor `mail_privacy.bats`  and adjusted `Makefile`, those could be extracted out to a separate PR? It's a few clicks for me to pull those commits out of the branch with cherry-pick into a stash that can be applied after this is merged, and push those up into a follow-up PR. Especially since they're cleanly isolated from the other changes in this PR.

georglauterbach commented:
> I also wanted to inquire about the changes that swap the `load` order to `setup,common` instead of `common,setup`? Was there a reason for that all? I had been using the latter as alphabetical sorting sweat_smile

No particular reason; I guess a matter of style again. I figured the setup functions are needed before the common functionality but it does not actually matter.

---

> If this is about the two commits that refactor `mail_privacy.bats` and adjusted `Makefile`, those could be extracted out to a separate PR? It's a few clicks for me to pull those commits out of the branch with cherry-pick into a stash that can be applied after this is merged, and push those up into a follow-up PR. Especially since they're cleanly isolated from the other changes in this PR.

Yes, this may be a good idea to keep the diff small. I will update this PR and apply PF feedback; you can then go ahead and pick them out of this PR :) We can then squash this PR, that's fine with me.

georglauterbach commented on test/helper/change-detection.bash:
```diff
@@ -1,29 +1,86 @@
+#
+# @param ${1} = expected count [OPTIONAL]
+# @param ${2} = container name [OPTIONAL]
+function _wait_until_expected_count_is_matched() {
```
I moved it here because as of now, it is only used here. If this ever changes, we can move it back.

georglauterbach commented:
UPDATE: Applied PR feedback; you can go ahead and pick out the 2 two commits as mentioned earlier @polarathene :)

polarathene commented:
LGTM :+1: 

I left your original two commits alone after extracting out the two files we discussed into their own PRs. Then merged the two review feedback commits into one.

casperklein commented on test/helper/common.bash:
```diff
@@ -1,247 +1,397 @@
   until "${@}"
   do
-    if [[ -n ${FATAL_FAILURE_TEST_COMMAND} ]] && ! eval "${FATAL_FAILURE_TEST_COMMAND}"; then
+    if [[ -n ${FATAL_FAILURE_TEST_COMMAND} ]] && ! eval "${FATAL_FAILURE_TEST_COMMAND}"
```
Whats the purpose of "eval" here, shouldn't it work flawless without too?

georglauterbach replied:
I thought that too, but remember that the whole command is just one string, not an array. Using `eval` here is similar to splitting the string at whitespaces and then using the array here.
